ast/ast.fbs

```diff
@@ -1,5 +1,64 @@
 namespace fbast;
 
+union MonoType {
+  NamedType,
+  ArrayType,
+  RecordType,
+  FunctionType,
+}
+
+table NamedType {
+  id:Identifier;
+}
+
+table ArrayType {
+  base_node:BaseNode;
+  element_type:MonoType;
+}
+
+table PropertyType {
+  base_node:BaseNode;
+  id:Identifier;
+  ty:MonoType;
+}
+
+table RecordType {
+  base_node:BaseNode;
+  properties:[PropertyType];
+  tvar:Identifier;
+}
+
+enum ParameterKind : byte {
+  Required,
+  Optional,
+  Pipe,
+}
+
+table ParameterType {
+  base_node:BaseNode;
+  name:Identifier;
+  ty:MonoType;
+  kind:ParameterKind;
+}
+
+table FunctionType {
+  base_node:BaseNode;
+  parameters:[ParameterType];
+  retn:MonoType;
+}
+
+table Constraint {
```

Comment:
Do we want to name this `TypeConstraint`? and maybe rename the Rust struct accordingly? Its a small thing but once we release it we can't easily change it.

Reply:
I agree.